runtime/src/iree/hal/drivers/hip/hip_driver.c

```diff
@@ -465,10 +472,120 @@ static iree_status_t iree_hal_hip_driver_create_device_by_index(
   IREE_HIP_RETURN_IF_ERROR(&driver->hip_symbols,
                            hipDeviceGet(&device, device_index), "hipDeviceGet");
 
-  iree_status_t status = iree_hal_hip_driver_create_device_by_id(
-      base_driver, IREE_HIPDEVICE_TO_DEVICE_ID(device), param_count, params,
-      host_allocator, out_device);
+  *out_id = IREE_HIPDEVICE_TO_DEVICE_ID(device);
+  return iree_ok_status();
+}
 
+static iree_status_t iree_hal_hip_driver_get_device_id_by_path(
+    iree_hal_driver_t* base_driver, iree_string_view_t device_path,
+    iree_hal_device_id_t* out_id) {
+  iree_hal_hip_driver_t* driver = iree_hal_hip_driver_cast(base_driver);
+  IREE_RETURN_IF_ERROR(iree_hal_hip_init(driver));
+
+  iree_hal_device_id_t device_id = 0;
+  iree_status_t status = iree_hal_hip_driver_get_device_id_by_uuid(
+      base_driver, device_path, &device_id);
+  if (iree_status_is_ok(status)) {
+    *out_id = device_id;
+    return iree_ok_status();
+  }
+  if (!iree_status_is_invalid_argument(status)) {
+    return status;
+  }
+  status = iree_hal_hip_driver_get_device_id_by_index(base_driver, device_path,
+                                                      &device_id);
+  if (iree_status_is_ok(status)) {
+    *out_id = device_id;
+    return iree_ok_status();
+  }
+  if (!iree_status_is_invalid_argument(status)) {
+    return status;
+  }
+  return iree_make_status(IREE_STATUS_UNIMPLEMENTED, "unsupported device path");
+}
+
+static iree_status_t iree_hal_hip_driver_create_multi_queue_device_by_ids(
+    iree_hal_driver_t* base_driver, iree_hal_device_id_t* device_ids,
+    iree_host_size_t device_count, iree_host_size_t param_count,
+    const iree_string_pair_t* params, iree_allocator_t host_allocator,
+    iree_hal_device_t** out_device) {
+  IREE_ASSERT_ARGUMENT(base_driver);
+  IREE_ASSERT_ARGUMENT(out_device);
+
+  iree_hal_hip_driver_t* driver = iree_hal_hip_driver_cast(base_driver);
+  IREE_TRACE_ZONE_BEGIN(z0);
```

Comment:
It's a good idea to factor your code to verify, allocate, populate, return - the most returns should happen during verification before you do anything you need to clean up, then if the allocation fails you can return safely as nothing has happened, populate should succeed, and then you return. If there's external side-effecting calls during population that you couldn't verify then you need to handle the cleanup of your allocated and other stateful things.

Here you should verify that device IDs are valid (so that check on 527 happens before you do anything). You malloc, populate the device IDs without the risk of failure, and try to call the inner method.

In this particular case you can also use the stack - since there's a practical device limit you can iree_alloca the device list space (by checking first that the device count is not over the limit) - we don't want to blow the stack but 512 bytes for 64 8byte pointers is fine and then you don't have to worry about cleaning up the allocation.